atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresGetCandidateCellsForSweeping.java

```diff
@@ -0,0 +1,326 @@
+//
+//   2. Limiting an outer query doesn't seem to work well, e.g.:
+//          SELECT ... FROM (
+//              SELECT ... FROM my_table
+//              ...
+//              ORDER BY row_name, col_name, ts
+//              ...
+//          ) sub
+//          WHERE ...
+//          ORDER BY row_name, col_name, ts
+//          LIMIT 1000
+//      Postgres will often choose a full table scan for the inner SELECT query, even when an index
+//      range scan is sufficient (because of the limit on the outer query). Postgres also doesn't
+//      have hints like Oracle, so one can't force it to use a certain plan. A trick with specifying
+//      a very large limit for the inner query seems to work, hinting the optimizer that only
+//      the first few rows matter:
+//          SELECT ... FROM (
+//              SELECT ... FROM my_table
+//              ...
+//              ORDER BY row_name, col_name, ts
+//              ...
+//              LIMIT 100000000000000000000 -- a very large number that is guaranteed to exceed the table size
+//          ) sub
+//          WHERE ...
+//          ORDER BY row_name, col_name, ts
+//          LIMIT 1000
+//      However, this looks very hacky and fragile, so instead we resort to simply limiting the inner query
+//      (i.e., limiting the number of examined cells rather than the number of returned candidates).
+//
+// So, our SQL query does the following:
+//
+//   1. Grab a page of SQL rows, starting at some (row_name, col_name, ts).
+//   2. Number the rows, starting with 1.
+//   3. Group the rows by cell key, i.e. (row_name, col_name), and compute a few aggregates:
+//          - Minimum and maximum row number
+//          - Minimum and maximum timestamp
+//          - An array of all timestamps (using ARRAY_AGG)
+//   4. If the THOROUGH strategy is being used, do a self-join to check if the values are empty.
+//   5. Filter out cells that don't satisfy the definition of a candidate, but keep the first
+//      and last SQL row in each page (except for the last page).
+//
+//
+public class PostgresGetCandidateCellsForSweeping implements DbKvsGetCandidateCellsForSweeping {
+
+    private final PostgresPrefixedTableNames prefixedTableNames;
+    private final SqlConnectionSupplier connectionPool;
+    private final IntToLongFunction sqlRowLimitProvider;
+
+    public static PostgresGetCandidateCellsForSweeping create(
+            PostgresPrefixedTableNames prefixedTableNames,
+            SqlConnectionSupplier connectionPool) {
+        return new PostgresGetCandidateCellsForSweeping(
+                prefixedTableNames,
+                connectionPool,
+                // Since in our SQL query we limit the number of examined cells rather than the number of
+                // returned candidates, we set the limit higher than the requested batch size.
+                batchHint -> 4 * batchHint);
```

Comment:
I understand what you are doing here, but this is strange and a bit counter-intuitive.  Why 4?  Maybe expose this as a pref?